src/test/java/uk/gov/hmcts/ccd/domain/service/caseaccess/CaseAccessOperationTest.java

```diff
@@ -85,6 +101,59 @@ void shouldHandleWrongJurisdiction() {
         }
     }
 
+    @Nested()
+    @DisplayName("updateUserAccess(reference, caseUser)")
+    class GrantAccessCaseUser {
+        private CaseDetails caseDetails;
+
+        @BeforeEach
+        void setUp() {
+            caseDetails = new CaseDetails();
+            caseDetails.setId(CASE_ID.toString());
+            caseDetails.setCaseTypeId(CASE_TYPE_ID);
+        }
+
+        @Test
+        @DisplayName("should reject update when it contains an unknown case role")
+        void shouldRejectWhenUnknownCaseRoles() {
+            final Executable execAccessUpdate = () -> caseAccessOperation.updateUserAccess(caseDetails,
+                                                                                           caseUser(NOT_CASE_ROLE));
+            assertThrows(InvalidCaseRoleException.class, execAccessUpdate);
+            verifyZeroInteractions(caseUserRepository);
+        }
+
+        @Test
+        @DisplayName("should grant access when added case role valid")
+        void shouldGrantAccessForCaseRole() {
+            caseAccessOperation.updateUserAccess(caseDetails, caseUser(CASE_ROLE));
+
+            verify(caseUserRepository).grantAccess(CASE_ID, USER_ID, CASE_ROLE);
+        }
+
+        @Test
+        @DisplayName("should revoke access for removed case roles")
+        void shouldRevokeRemovedCaseRoles() {
+            caseAccessOperation.updateUserAccess(caseDetails, caseUser(CASE_ROLE));
+
+            verify(caseUserRepository).revokeAccess(CASE_ID, USER_ID, CASE_ROLE_GRANTED);
+        }
+
+        @Test
+        @DisplayName("should ignore case roles already granted")
+        void shouldIgnoreGrantedCaseRoles() {
+            caseAccessOperation.updateUserAccess(caseDetails, caseUser(CASE_ROLE_GRANTED));
+
+            verify(caseUserRepository, never()).grantAccess(CASE_ID, USER_ID, CASE_ROLE_GRANTED);
+        }
+    }
+
```

Comment:
Are we missing a case where there is an update without case roles?
```
Given a case and a user
When granting access to the case for the user ID WITHOUT case role
Then 400 Case Role mandatory 
And user is not granted access
```

Reply:
Same for a case where there is invalid case id?

```
Given an invalid case ID and a user
When granting access to the case for the user ID with a valid case role
Then 400 invalid case ID
And user is not granted access
```

Reply:
I guess the above + few other ACs are checked on another level before this logic.

Reply:
High-level input validation, like ID checksum, are checked one level above in subsequent PR #308 . However, I've actually changed the implementation to accept update without case roles, as I felt this was needed for `PUT` to keep its true meaning: PUT without roles => remove all accesses. Otherwise, this would require `POST` + `DELETE` and it doesn't feel as nice.
But I did forget to update the ticket, sorry for that 🙇

Reply:
Ok makes sense, thanks.